src/runtime_src/core/pcie/linux/device_linux.cpp

```diff
@@ -1030,6 +1031,35 @@ struct sysfs_fcn<std::vector<VectorValueType>>
   }
 };
 
+struct memory_topology
+{
+  using data_type = query::mem_topology::data_type;
+  using result_type = query::mem_topology::result_type;
+
+  static result_type
+  get(const xrt_core::device* device, key_type)
+  {
+    const auto data = xrt_core::device_query<query::mem_topology_raw>(device);
+    const auto mem_topo = reinterpret_cast<const struct mem_topology*>(data.data());
+    const auto xclbin_uuid = xrt_core::device_query<query::xclbin_uuid>(device);
+    result_type topology;
+    for (int32_t index = 0; index < mem_topo->m_count; index++) {
+      data_type mem_data;
+      mem_data.xclbin_uuid = xclbin_uuid;
+      mem_data.hw_context_slot = 0;
+      mem_data.m_type = mem_topo->m_mem_data[index].m_type;
+      mem_data.m_used = mem_topo->m_mem_data[index].m_used;
+      memcpy(mem_data.m_tag, mem_topo->m_mem_data[index].m_tag, sizeof(mem_topo->m_mem_data[index].m_tag));
+      // The following two entries are unions
+      // Within the union using any name would work. So use the first!
+      mem_data.m_size = mem_topo->m_mem_data[index].m_size;
+      mem_data.m_base_address = mem_topo->m_mem_data[index].m_base_address;
+      topology.push_back(mem_data);
+    }
+    return topology;
```

Comment:
This is not really a query request, it is basically just doing formatting.   It does not belong here.